src/inline-assembly.md

```diff
@@ -463,6 +463,7 @@ To avoid undefined behavior, these rules must be followed when using function-sc
 - The compiler cannot assume that the instructions in the asm are the ones that will actually end up executed.
   - This effectively means that the compiler must treat the `asm!` as a black box and only take the interface specification into account, not the instructions themselves.
   - Runtime code patching is allowed, via target-specific mechanisms.
+  - However there is no guarantee that each `asm!` directly corresponds to a single instance of instructions in the object file: the compiler is free to duplicate or deduplicate `asm!` blocks.
```

Comment:
I think this should explicitly specify the other direction as well (no guarantee that an instance of instructions in the object file corresponds to exactly one `asm!` expression). When only stated in the current direction, the lack of guarantee could still be understood as only permitting duplication or removal of `asm!`, where unification is the originating observed behavior.

Separately, it would be good to specify when disjoint `asm!` are able to be unified, given the prior assertion that the compiler is not permitted to rely on the contents of the `asm!` string being the instructions which get executed. I think the correct precondition would be that the instantiated `asm!` strings are *structurally equivalent* — that the directives/instructions given to the assembler are identical.

The compiler is not permitted to interpret the *effect* of the directives/instructions, but it *is* permitted to inspect the *identity* of thus.

Reply:
> I think this should explicitly specify the other direction as well (no guarantee that an instance of instructions in the object file corresponds to exactly one `asm!` expression). When only stated in the current direction, the lack of guarantee could still be understood as only permitting duplication or removal of `asm!`, where unification is the originating observed behavior.

I think this is already covered by the word "deduplicate"?



> Separately, it would be good to specify when disjoint `asm!` are able to be unified, given the prior assertion that the compiler is not permitted to rely on the contents of the `asm!` string being the instructions which get executed. I think the correct precondition would be that the instantiated `asm!` strings are _structurally equivalent_ — that the directives/instructions given to the assembler are identical.
> 
> The compiler is not permitted to interpret the _effect_ of the directives/instructions, but it _is_ permitted to inspect the _identity_ of thus.

I think that we may have to allow more flexibility here due to things like link-time identical code folding. This would merge separate asm that may be structurally different but end up emitting the same bytes. I'm open to suggestions on the exact wording to use here.